File: hw/chalan/u3/hw28PrinciplesAplication/FRUITS/JS/fruit-store/src/controller/FruitController.js
import MongoDBUtil from '../utils/MongoDBUtil.js'
import Fruit from '../model/Fruit.js'

class FruitController {
  async init() {
    const db = await MongoDBUtil.getDatabase()
    this.collection = db.collection('fruits')
  }

  async addFruit(name, price, stock) {
    await this.collection.insertOne(new Fruit(name, price, stock))
  }

  async deleteFruit(name) {
    await this.collection.deleteOne({ name })
  }

  async buyFruit(name, quantity) {
    const fruit = await this.collection.findOne({ name })
    if (!fruit) return -1
    if (quantity > fruit.stock) return -2
    await this.collection.updateOne(
      { name },
      { $set: { stock: fruit.stock - quantity } }
    )
    return fruit.price * quantity
  }

  async getFruitNames() {
    return (await this.collection.find().toArray()).map(f => f.name)
  }

  async getPriceByName(name) {
    const fruit = await this.collection.findOne({ name })
    return fruit ? fruit.price : 0
  }

  async existsFruit(name) {
    return !!(await this.collection.findOne({ name }))
  }
}

export default FruitController

File: hw/chalan/u3/hw26GUIs/FRUITS/Python/storefruit/controller/fruit_controller.py
from pymongo import MongoClient
from utils.mongodb_util import MongoDBUtil


class FruitController:
    def __init__(self):
        db = MongoDBUtil.get_database()
        self.collection = db["fruits"]

    # ===== CRUD =====

    def add_fruit(self, name, price, stock):
        self.collection.insert_one({
            "name": name,
            "price": price,
            "stock": stock
        })

    def delete_fruit(self, name):
        self.collection.delete_one({"name": name})

    # ===== BUSINESS =====

    def buy_fruit(self, name, quantity):
        fruit = self.collection.find_one({"name": name})

        if fruit is None:
            return -1

        if quantity > fruit["stock"]:
            return -2

        new_stock = fruit["stock"] - quantity
        self.collection.update_one(
            {"name": name},
            {"$set": {"stock": new_stock}}
        )

        return fruit["price"] * quantity

    # ===== QUERIES =====

    def get_fruit_names(self):
        return [f["name"] for f in self.collection.find()]

    def get_price_by_name(self, name):
        fruit = self.collection.find_one({"name": name})
        if fruit is None:
            return 0
        return fruit["price"]

    def get_stock_by_name(self, name):
        fruit = self.collection.find_one({"name": name})
        if fruit is None:
            return 0
        return fruit["stock"]

    def exists_fruit(self, name):
        return self.collection.find_one({"name": name}) is not None
